include/aspect/particle/particle_accessor.h

```diff
@@ -0,0 +1,188 @@
+/*
+ Copyright (C) 2017 by the authors of the ASPECT code.
+
+ This file is part of ASPECT.
+
+ ASPECT is free software; you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation; either version 2, or (at your option)
+ any later version.
+
+ ASPECT is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with ASPECT; see the file LICENSE.  If not see
+ <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _aspect_particle_particle_accessor_h
+#define _aspect_particle_particle_accessor_h
+
+#include <aspect/global.h>
+#include <aspect/particle/particle.h>
+
+#include <deal.II/base/array_view.h>
+
+namespace aspect
+{
+  namespace Particle
+  {
+    using namespace dealii;
+
+    template <int, int> class ParticleIterator;
+
+    template<int dim, int spacedim=dim>
+    class ParticleAccessor
+    {
+      private:
+        /**
+         * A pointer to the container that stores the particles. Obviously,
+         * this accessor is invalidated if the container changes.
+         */
+        std::multimap<types::LevelInd, Particle<dim,spacedim> >            *map;
+
+        /**
+         * An iterator into the container of particles. Obviously,
+           * this accessor is invalidated if the container changes.
+         */
+        typename std::multimap<types::LevelInd, Particle<dim,spacedim> >::iterator  particle;
+
+        /**
+         * Make ParticleIterator a friend to allow it constructing ParticleAccessors.
+         */
+        template <int, int> friend class ParticleIterator;
```

Comment:
I think we always move the friend declarations to the very end of the class. Can you do the same here?